featurebyte/query_graph/node/function.py

```diff
@@ -0,0 +1,227 @@
+from typing import Any, Dict, List, Literal, Optional, Sequence, Tuple, Union
+from typing_extensions import Annotated
+
+from abc import abstractmethod  # pylint: disable=wrong-import-order
+
+from pydantic import BaseModel, Field
+
+from featurebyte.enum import DBVarType, FunctionParameterInputForm
+from featurebyte.models.base import PydanticObjectId
+from featurebyte.query_graph.enum import NodeOutputType, NodeType
+from featurebyte.query_graph.node.base import BaseSeriesOutputNode
+from featurebyte.query_graph.node.metadata.operation import (
+    DerivedDataColumn,
+    NodeOutputCategory,
+    OperationStructure,
+    OperationStructureBranchState,
+    OperationStructureInfo,
+    PostAggregationColumn,
+)
+from featurebyte.query_graph.node.metadata.sdk_code import (
+    CodeGenerationConfig,
+    CodeGenerationContext,
+    ExpressionStr,
+    InfoDict,
+    RightHandSide,
+    StatementT,
+    ValueStr,
+    VariableNameGenerator,
+    VarNameExpressionInfo,
+)
+
+
+class BaseFunctionParameterInput(BaseModel):
+    """BaseFunctionParameterInput class"""
+
+    dtype: DBVarType
+
+    @abstractmethod
+    def get_column_args(self) -> List[Optional[str]]:
+        """
+        Get column arguments (column name or None) for the function parameter input.
+        If the input is a value, return empty list.
+
+        Returns
+        _______
+        List[Optional[str]]
+        """
+
+    @abstractmethod
+    def get_sdk_function_argument(
+        self, node_inputs: List[VarNameExpressionInfo], node_input_index: int
+    ) -> Tuple[RightHandSide, int]:
+        """
+        Get SDK function argument for the function parameter input.
+
+        Parameters
+        ----------
+        node_inputs : List[VarNameExpressionInfo]
+            List of node inputs.
+        node_input_index : int
+            The index of the node input that the function parameter input is associated with.
+
+        Returns
+        -------
+        Tuple[RightHandSide, int]
+            Tuple of SDK function argument and the updated node input index.
+        """
+
+
+class ValueFunctionParameterInput(BaseFunctionParameterInput):
+    """ValueFunctionParameterInput class"""
+
+    value: Optional[Any]
+    input_form: Literal[FunctionParameterInputForm.VALUE] = Field(
+        FunctionParameterInputForm.VALUE, const=True
+    )
+
+    def get_column_args(self) -> List[Optional[str]]:
+        return []
+
+    def get_sdk_function_argument(
+        self, node_inputs: List[VarNameExpressionInfo], node_input_index: int
+    ) -> Tuple[RightHandSide, int]:
+        # do not increment node_input_index as this is a value input, it should not increment the node_input_index
+        return ValueStr.create(self.value), node_input_index
+
+
+class ColumnFunctionParameterInput(BaseFunctionParameterInput):
+    """ColumnFunctionParameterInput class"""
+
+    column_name: Optional[str]
```

Comment:
Is the `column_name` optional because the input can be a temporary column that is not assigned into a view? If the input is a feature, will this refer to the feature name?

Reply:
Yes, `column_name` because it can be a derived column if it is not assigned to a view. If the input is a feature, this will refer to the feature name.

Reply:
The value of `column_name` is not used in SDK feature derivation. The main reason to keep this is to support `get_required_input_columns` method (https://github.com/featurebyte/featurebyte/blob/cc24823e663e72e1a9369785d33950c637ce57cd/featurebyte/query_graph/transform/reconstruction.py#L167-L172).